images/cfw/opt/x1plus/lib/python/x1plus/services/x1plusd/__init__.py

```diff
@@ -15,44 +18,94 @@
-        self.httpd = HTTPService(router=self.router, daemon=self)
-        self.sensors = SensorsService(router=self.router, daemon=self)
-        self.gpios = GpioManager(daemon=self)
-        self.expansion = ExpansionManager(router=self.router, daemon=self)
+        self.httpd = HTTPService(daemon=self)
+        self.sensors = SensorsService(daemon=self)
         self.mcproto = MCProtoParser(daemon=self)
-        self.actions = ActionHandler(router=self.router, daemon=self)
-        if not self.settings.get("polar_cloud", False):
-            self.polar_cloud = None
-        else:
-            from .polar_cloud import PolarPrintService
-            self.polar_cloud = PolarPrintService(daemon=self)
+        self.expansion = ExpansionManager(daemon=self)
+        self.gpios = GpioManager(daemon=self)
+        self.actions = ActionHandler(daemon=self)
         
+        self.custom_modules = {}
+        """
+        { "name": ClassInstance }
+        """
+
+        self.watched_keys = []
+        self.MODULES = []
+
+        BASE_MODULE_DIR = os.path.abspath(os.path.join(os.path.dirname(os.path.abspath(__file__)), "modules"))
+        self.load_modules(BASE_MODULE_DIR)
+
+        SDCARD_MODULE_DIR = "/mnt/sdcard/x1plus/modules"
+        if not os.path.isdir(SDCARD_MODULE_DIR):
+            os.makedirs(SDCARD_MODULE_DIR)
+
+        # Custom modules are loaded after all core modules
+        # Full import paths will work if a dependency exists
+        self.load_modules(SDCARD_MODULE_DIR)
+
         return self
 
+    def load_modules(self, module_directory: str):
+        self.MODULES.extend(module_importer(module_directory, include_subdirs=True))
+
+        for scanned_module in self.MODULES:
+            key = scanned_module.get('key')
+            module = scanned_module.get('module', None)
+            
+            if key not in self.watched_keys:
+                self.watched_keys.append(key)
+
+            enabled = scanned_module.get("config", {}).get("enabled", "").lower() == "true"
+            if module and self.settings.get(key, enabled) and hasattr(module, "load"):
+                name = scanned_module.get("name")
+                try:
+                    module.load(daemon=self)
+                except Exception as e:
+                    logger.error(f"Error loading x1plusd module {name}. {e.__class__.__name__}: {e}")
+
+
+
     async def start(self):
-        asyncio.create_task(self.mqtt.task())
         asyncio.create_task(self.settings.task())
+
+        asyncio.create_task(self.mqtt.task())
         asyncio.create_task(self.ota.task())
         asyncio.create_task(self.httpd.task())
         asyncio.create_task(self.sensors.task())
         asyncio.create_task(self.expansion.task())
         asyncio.create_task(self.mcproto.task())
         asyncio.create_task(self.actions.task())
-        if self.polar_cloud:
-            asyncio.create_task(self.polar_cloud.begin())
+
+        for scanned_module in self.MODULES:
```

Comment:
for that point, perhaps this ought be split into modules_available and modules_enabled -- and then we don't have to go through the settings dance here, we just have to iterate through the modules_enabled

Reply:
I like this, would also let us poll for it later if we wanted a quick glance as what's enabled or not without needing to iterate and check.